File: 14/b.ts
// const input = await Bun.file("test.txt").text();
const input = await Bun.file("input.txt").text();

let res = 0;

const lines = input.split("\n").map((line) => line.split(""));

const prevStates: string[] = [];

for (let c = 0; c < 1000000000; c++) {
  if (c % 1_000_000 === 0) console.log(c);
  for (let j = 0; j < lines[0].length; j++) {
    for (let i = 0; i < lines.length; i++) {
      if (lines[i][j] !== ".") continue;
      for (let k = i + 1; k < lines.length; k++) {
        if (lines[k][j] === "#") break;
        if (lines[k][j] === "O") {
          lines[k][j] = ".";
          lines[i][j] = "O";
          i++;
        }
      }
    }
  }
  for (let i = 0; i < lines.length; i++) {
    for (let j = 0; j < lines[i].length; j++) {
      if (lines[i][j] !== ".") continue;
      for (let k = j + 1; k < lines[i].length; k++) {
        if (lines[i][k] === "#") break;
        if (lines[i][k] === "O") {
          lines[i][k] = ".";
          lines[i][j] = "O";
          j++;
        }
      }
    }
  }
  for (let j = 0; j < lines[0].length; j++) {
    for (let i = lines.length - 1; i >= 0; i--) {
      if (lines[i][j] !== ".") continue;
      for (let k = i - 1; k >= 0; k--) {
        if (lines[k][j] === "#") break;
        if (lines[k][j] === "O") {
          lines[k][j] = ".";
          lines[i][j] = "O";
          i--;
        }
      }
    }
  }
  for (let i = 0; i < lines.length; i++) {
    for (let j = lines[i].length - 1; j >= 0; j--) {
      if (lines[i][j] !== ".") continue;
      for (let k = j - 1; k >= 0; k--) {
        if (lines[i][k] === "#") break;
        if (lines[i][k] === "O") {
          lines[i][k] = ".";
          lines[i][j] = "O";
          j--;
        }
      }
    }
  }

  let str = JSON.stringify(lines);
  let index = prevStates.indexOf(str);
  if (index !== -1) {
    console.log("Found loop", index, c);
    c += Math.floor((1000000000 - c) / (c - index)) * (c - index);
  }
  prevStates.push(str);
}

console.log(lines.map((line) => line.join("")).join("\n"));

lines.reverse();

for (let i = 0; i < lines.length; i++) {
  for (let j = 0; j < lines[i].length; j++) {
    if (lines[i][j] === "O") res += i + 1;
  }
}

console.log(res);

File: 14/b.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {
	file, _ := os.ReadFile("input.txt")

	rawLines := strings.Split(string(file), "\n")
	lines := make([][]rune, len(rawLines))
	for i, line := range rawLines {
		lines[i] = []rune(line)
	}
	previous := make([][][]rune, 0, 0)
	for c := 0; c < 1000000000; c++ {
		if c%100000 == 0 {
			fmt.Println(c)
		}
		for j := 0; j < len(lines[0]); j++ {
			for i := 0; i < len(lines); i++ {
				if lines[i][j] != '.' {
					continue
				}
				for k := i + 1; k < len(lines); k++ {
					if lines[k][j] == '#' {
						break
					}
					if lines[k][j] == 'O' {
						lines[k][j] = '.'
						lines[i][j] = 'O'
						i++
					}
				}
			}
		}
		for i := 0; i < len(lines); i++ {
			for j := 0; j < len(lines[0]); j++ {
				if lines[i][j] != '.' {
					continue
				}
				for k := j + 1; k < len(lines); k++ {
					if lines[i][k] == '#' {
						break
					}
					if lines[i][k] == 'O' {
						lines[i][k] = '.'
						lines[i][j] = 'O'
						j++
					}
				}
			}
		}
		for j := 0; j < len(lines[0]); j++ {
			for i := len(lines) - 1; i >= 0; i-- {
				if lines[i][j] != '.' {
					continue
				}
				for k := i - 1; k >= 0; k-- {
					if lines[k][j] == '#' {
						break
					}
					if lines[k][j] == 'O' {
						lines[k][j] = '.'
						lines[i][j] = 'O'
						i--
					}
				}
			}
		}
		for i := 0; i < len(lines); i++ {
			for j := len(lines[0]) - 1; j >= 0; j-- {
				if lines[i][j] != '.' {
					continue
				}
				for k := j - 1; k >= 0; k-- {
					if lines[i][k] == '#' {
						break
					}
					if lines[i][k] == 'O' {
						lines[i][k] = '.'
						lines[i][j] = 'O'
						j--
					}
				}
			}
		}
		for i, prev := range previous {
			isSame := true
			for j := 0; j < len(prev); j++ {
				for k := 0; k < len(previous[i][j]); k++ {
					if previous[i][j][k] != lines[j][k] {
						isSame = false
						break
					}
				}
			}
			if isSame {
				fmt.Println("Found", i, c)
				c += (1000000000 - c) / (c - i) * (c - i)
				break
			}
		}
		newLines := make([][]rune, len(lines))
		for i := range lines {
			newLines[i] = make([]rune, len(lines[i]))
			copy(newLines[i], lines[i])
		}
		previous = append(previous, newLines)
	}

	res := 0
	for i, line := range lines {
		for _, char := range line {
			fmt.Print(string(char))
			if char == 'O' {
				res += len(lines) - i
			}
		}
		fmt.Println()
	}

	fmt.Println()

	fmt.Println(res)
}
